more/pixels/gallery.js

```diff
@@ -81,230 +55,168 @@ export function makeGallery(
-        pixelListPayment,
-      );
-      await gallerySplitPixelPurse.depositAll(exclusivePayment); // conserve pixels
-
-      const { transferAmount, useAmount } = pixelAssay.toTransferAndUseRights(
-        pixelListAmount,
-        useRightAssay,
-        transferRightAssay,
-      );
-
-      const transferRightPurse = transferRightMint.mint(transferAmount);
-      const useRightPurse = useRightMint.mint(useAmount);
+  function setPixelListState(pixelList, newColor) {
+    for (let i = 0; i < pixelList.length; i += 1) {
+      const pixel = pixelList[i];
+      state[pixel.x][pixel.y] = newColor;
+      // eslint-disable-next-line no-use-before-define
+      lruQueue.requeue(pixel);
+    }
 
-      const transferRightPayment = await transferRightPurse.withdrawAll(
-        'transferRights',
-      );
-      const useRightPayment = await useRightPurse.withdrawAll('useRights');
+    // for now we pass the whole state, but only once per call
+    stateChangeHandler(getState());
+  }
 
-      return {
-        transferRightPayment,
-        useRightPayment,
-      };
-    });
+  // anyone can getPixelColor, no restrictions, no tokens
+  function getPixelColor(x, y) {
+    const rawPixel = harden({ x: Nat(x), y: Nat(y) });
+    return state[rawPixel.x][rawPixel.y];
   }
 
-  // merge UseRights and TransferRights into a pixel
-  async function toPixel(transferRightPaymentP) {
-    return Promise.resolve(transferRightPaymentP).then(
-      async transferRightPayment => {
-        // someone else may have the useRightPayment so we must destroy the
-        // useRight
-
-        // we have an exclusive on the transfer right
-        const transferAmount = transferRightPayment.getBalance();
-        await transferRightIssuer.getExclusiveAll(transferRightPayment);
-
-        const pixelListAmount = transferRightAssay.toPixel(
-          transferAmount,
-          pixelAssay,
-        );
-
-        const { useAmount } = pixelAssay.toTransferAndUseRights(
-          pixelListAmount,
-          useRightAssay,
-          transferRightAssay,
-        );
-
-        // commit point
-        await useRightMint.destroy(useAmount);
-        await transferRightMint.destroy(transferAmount);
-
-        const pixelPayment = await gallerySplitPixelPurse.withdraw(
-          pixelListAmount,
-          'pixels',
-        ); // conserve pixels
-        return pixelPayment;
+  // makeUseObj is part of the configuration passed into makeMint and
+  // is used to create the "use object" that is associated with an
+  // underlying asset (purse or payment). In this case, the use object
+  // has the methods for changing the color of pixels
+
+  function makeUseObj(issuer, asset) {
+    const useObj = harden({
+      // change the color of the pixels in the amount after checking
+      // that the asset has the authority to do so.
+      changeColor(amount, newColor) {
+        insistNonEmptyAmount(issuer, amount);
```

Comment:
I am disturbed by the non-uniform handling of the zero case. If the amount passed in is empty, why not honor it by doing nothing?

Admittedly, this may be too hazard prone in the current design, and this non-uniform handling of the zero case may catch more bugs than it causes. But, in the upcoming linear design, assuming we distinguish zero from used up, we can avoid honoring even a zero amount on a used up asset without a special case. Until then, we should at least have a todo to revisit it at that time.

Reply:
I've added a todo to look into allowing empty amounts to be colored without throwing